### PR TITLE
Preserve trailing newline in json

### LINKED ISSUE #209: Keep trailing newline in JSON
We can't [keep](https://github.com/flathub/flatpak-external-data-checker/issues/192) most of the JSON formatting, but we could at least preserve trailing newline in the manifest, i.e. add it manually if it was there before we dump updated manifest.

### PR DESCRIPTION
**Pull request by cpba**
Preserve the trailing newline in json manifests by adding it
back if it was present before dumping the updated manifest.

Fixes #209

### CODE REVIEW

**Comment by cpba** [comment]
Running the unit tests gives me back:
`FAILED (failures=1, expected failures=3)`

But I don't quite understand what I broke.

File: src/lib/utils.py [hunk 1]
```diff
@@ -425,13 +433,17 @@ def dump_manifest(contents, manifest_path):
     """Writes back 'contents' to 'manifest_path'.
 
     For YAML, we make a best-effort attempt to preserve
-    formatting; for JSON, we use the canonical 4-space indentation."""
+    formatting; for JSON, we use the canonical 4-space indentation,
+    but add a trailing newline if originally present."""
     _, ext = os.path.splitext(manifest_path)
+    newline = _check_newline(manifest_path)
     with open(manifest_path, "w", encoding="utf-8") as fp:
```
Comment (gasinvein) [inline src/lib/utils.py]: It would be a bit better if we could avoid opening the output file twice. Maybe open it in read-write mode here, and pass `fp` to `_check_newline`? Just make sure to `seek` the file to proper positions when using read-write mode.

File: src/lib/utils.py [hunk 2]
```diff
@@ -68,6 +68,14 @@ def _extract_timestamp(info):
     raise CheckerRemoteError(f"Cannot parse date/time: {date_str}")
 
 
+def _check_newline(manifest_path):
+    with open(manifest_path, "r") as f:
+        if f.read()[-1] == "\n":
```
Comment (gasinvein) [inline src/lib/utils.py]: This could be done in a less expensive way by seeking the file to the end and reading only the last char.

**Review by gasinvein — COMMENTED** [review]
Here a some performance-related suggestions. 

Also please add a test case for this.

File: src/lib/utils.py (hunk 1 above)
Comment (cpba) [reply src/lib/utils.py]: Sounds good, I'll change it and push a new commit.

**Comment by cpba** [comment]
I believe the unit test here fails because it expects to get back a file without the trailing newline:
https://github.com/flathub/flatpak-external-data-checker/blob/master/tests/test_checker.py#L395

Should I modify it too?

**Comment by gasinvein** [comment]
> I believe the unit test here fails because it expects to get back a file without the trailing newline: https://github.com/flathub/flatpak-external-data-checker/blob/master/tests/test_checker.py#L395
> 
> Should I modify it too?

Yeah, that should do.

**Comment by cpba** [comment]
I forgot to run black again after the changes

File: src/lib/utils.py [hunk 3]
```diff
@@ -68,6 +68,15 @@ def _extract_timestamp(info):
     raise CheckerRemoteError(f"Cannot parse date/time: {date_str}")
 
 
+def _check_newline(fp):
+    fp.seek(0, 2)
+    fp.seek(fp.tell()-1,0)
```
Comment (gasinvein) [inline src/lib/utils.py]: Please use named constants instead of numeric values, e.g. `os.SEEK_END` instead of `2`.
Also I guess this can be done in one action with negative offset instead of two, i.e. `fp.seek(-1, os.SEEK_END)`?

File: src/lib/utils.py [hunk 4]
```diff
@@ -425,13 +434,19 @@ def dump_manifest(contents, manifest_path):
...
+    formatting; for JSON, we use the canonical 4-space indentation,
+    but add a trailing newline if originally present."""
     _, ext = os.path.splitext(manifest_path)
-    with open(manifest_path, "w", encoding="utf-8") as fp:
+    with open(manifest_path, "r+", encoding="utf-8") as fp:
         if ext in (".yaml", ".yml"):
             _yaml.dump(contents, fp)
         else:
+            newline = _check_newline(fp)
+            fp.seek(0, 0)
```
Comment (gasinvein) [inline src/lib/utils.py]: Nitpick: remember and return to the original position in `_check_newline` instead resetting it outside.

File: tests/test_util.py [hunk 5]
```diff
@@ -222,5 +223,14 @@ async def test_wrong_content_type(self):
             )
 
 
+class TestCheckNewline(unittest.TestCase):
+    def test_newline(self):
+        fp = open("tests/com.jetbrains.PhpStorm.json", "r+", encoding="utf-8")
+        self.assertEqual(_check_newline(fp), True)
```
Comment (gasinvein) [inline tests/test_util.py]: Ideally, we should do a full "newline preservation" test, i.e. create a dummy `.json` file with newline, read it, dump it, check that the newline is there, and do the same for a file without trailing newline.

File: src/lib/utils.py [hunk 6]
```diff
@@ -68,6 +68,15 @@ def _extract_timestamp(info):
     raise CheckerRemoteError(f"Cannot parse date/time: {date_str}")
 
 
+def _check_newline(fp):
+    fp.seek(0, 2)
+    fp.seek(fp.tell() - 1, 0)
```
Comment (gasinvein) [inline src/lib/utils.py]: Please use named constants instead of numeric values, e.g. `os.SEEK_END` instead of `2`.
Also I guess this can be done in one action with negative offset instead of two, i.e. `fp.seek(-1, os.SEEK_END)`?

Comment (cpba) [reply src/lib/utils.py]: > Please use named constants instead of numeric values, e.g. os.SEEK_END instead of 2

Will do.

> Also I guess this can be done in one action with negative offset instead of two, i.e. fp.seek(-1, os.SEEK_END)?

Technically yes, but it requires opening the file in binary mode and I was afraid of edge cases with different encodings. Should I change it?

File: src/lib/utils.py (hunk 4 above)
Comment (cpba) [reply src/lib/utils.py]: I agree, that would be less ugly.

File: tests/test_util.py (hunk 5 above)
Comment (cpba) [reply tests/test_util.py]: If I'm reading this correctly, should I add new json files specifically for this test rather than reuse the ones I did?

And the test should also check whether the json file is correctly generated with the trailing newline, rather than relying only on the existing test at https://github.com/flathub/flatpak-external-data-checker/blob/master/tests/test_checker.py#L141?

File: src/lib/utils.py (hunk 6 above)
Comment (gasinvein) [reply src/lib/utils.py]: Right. I'm not sure if `json.dump` will play nice with binary IO, so it's probably safer to leave it as is, then.

File: tests/test_util.py (hunk 5 above)
Comment (gasinvein) [reply tests/test_util.py]: > should I add new json files specifically for this test rather than reuse the ones I did?

Yes, although you can embed the files contents into the test code instead of storing as real files in git.

>And the test should also check whether the json file is correctly generated with the trailing newline

Not necessary, but probably it would be even better.

So the test flow will look something like this:
1. Check if `_check_newline` works properly on a file with known contents
2. Read and dump again the file
3. Use `_check_newline` to check if trailing newline was preserved

**Comment by cpba** [comment]
Alright, I think it's looking more presentable now.

File: src/lib/utils.py [hunk 7]
```diff
@@ -425,13 +436,18 @@ def dump_manifest(contents, manifest_path):
...
-    with open(manifest_path, "w", encoding="utf-8") as fp:
+    with open(manifest_path, "r+", encoding="utf-8") as fp:
         if ext in (".yaml", ".yml"):
             _yaml.dump(contents, fp)
         else:
+            newline = _check_newline(fp)
             json.dump(obj=contents, fp=fp, indent=4)
+            if newline:
+                fp.write("\n")
+        fp.truncate()
```
Comment (gasinvein) [inline src/lib/utils.py]: I'm not sure I understand why this is needed.

File: src/lib/utils.py [hunk 8]
```diff
@@ -68,6 +68,17 @@ def _extract_timestamp(info):
     raise CheckerRemoteError(f"Cannot parse date/time: {date_str}")
 
 
+def _check_newline(fp):
+    fp.seek(0, os.SEEK_END)
+    fp.seek(fp.tell() - 1, os.SEEK_SET)
+    last_char = fp.read()
+    fp.seek(0, os.SEEK_SET)
```
Comment (gasinvein) [inline src/lib/utils.py]: Nitpick: It would nicer to remember the actual initial position and reset to it after we're done, instead of resetting to the beginning.

File: tests/test_newlinepreservation.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+
+from src.lib.utils import (
+    _check_newline,
+    dump_manifest,
+    read_manifest,
+)
+
+
+MANIFEST_WITH_NEWLINE = """{
+  "app-id": "com.gitlab.coringao.cavestory-nx",
```
Comment (gasinvein) [inline tests/test_newlinepreservation.py]: We don't need an actual flatpak manifest here to test dict dumping. Any json would do, even something like `{"one": 1}`.

File: src/lib/utils.py (hunk 7 above)
Comment (cpba) [reply src/lib/utils.py]: I thought there might potentially be cases where the new file is shorter than the original, since we're using `r+` mode we could end up with garbage at the end of the file if we don't truncate.

File: tests/test_newlinepreservation.py (hunk 9 above)
Comment (cpba) [reply tests/test_newlinepreservation.py]: It's one of mine, I just had it lying around and threw it in for realism. Would it be better to replace it with a minimal one liner?

File: src/lib/utils.py (hunk 8 above)
Comment (cpba) [reply src/lib/utils.py]: Yeah that'd be safer.

File: src/lib/utils.py (hunk 7 above)
Comment (gasinvein) [reply src/lib/utils.py]: Ah, indeed, good catch. I didn't think about it when suggested to use read-write mode. Now opening opening the file twice - in read-only and write-only modes - seems like a better idea :-/

File: tests/test_newlinepreservation.py (hunk 9 above)
Comment (gasinvein) [reply tests/test_newlinepreservation.py]: I think it will, yeah. No need to add more text than necessary.

File: src/lib/utils.py (hunk 7 above)
Comment (cpba) [reply src/lib/utils.py]: Do you want me to go back to the original implementation?

Comment (gasinvein) [reply src/lib/utils.py]: Yeah, if you don't mind, sorry for the confusion. Just don't change `_check_newline` argument back to a file path - instead, open the file for reading and detecting trailing newline here, in `dump_manifest` (maybe better outside this `with open()` block).

Comment (cpba) [reply src/lib/utils.py]: So if I understand this right, you want me to keep `_check_newline` as it currently is but just change `dump_manifest` back to `w` mode and open it for reading beforehand? I'll send a new commit and see if that's what you want.

Comment (gasinvein) [reply src/lib/utils.py]: Yep, something like this.

**Review by gasinvein — APPROVED** [review]

**Comment by gasinvein** [comment]
@cpba Thanks for your contribution. Don't mind if I squash your commits, since most of them redid previous ones?

**Comment by cpba** [comment]
No, I don't mind if you squash them, otherwise I'd have to go clean up my
commit messages :)

**Comment by gasinvein** [comment]
Ouch. Looks like Mypy was updated and now complains about ruamel.yaml, failing our type check test. I guess it's unrelated to this PR.
